boards/rp-pico/examples/pico_interpolator.rs

```diff
@@ -0,0 +1,441 @@
+    interp.get_lane1().set_base(255);
+    let values: [(i32, i32); 9] = [
+        (-1024, 0),
+        (-768, 0),
+        (-512, 0),
+        (-256, 0),
+        (0, 0),
+        (256, 64),
+        (512, 128),
+        (768, 192),
+        (1024, 255),
+    ];
+    for (arg, result) in values {
+        interp.get_lane0().set_accum(arg as u32);
+        if result != interp.get_lane0().peek() as i32 {
+            return false;
+        }
+    }
+    true
+}
+
+fn texture_mapping(interp: &mut Interp0) -> bool {
+    #[rustfmt::skip]
+    let texture: [u8;16] = [
+        0x00, 0x01, 0x02, 0x03,
+        0x10, 0x11, 0x12, 0x13,
+        0x20, 0x21, 0x22, 0x23,
+        0x30, 0x31, 0x32, 0x33,
+    ];
+
+    // the position will be given in fixed point with 16 bits
+    // fractional part
+    let uv_fractional_bits = 16;
+    let texture_width_bits = 2;
+    let texture_height_bits = 2;
+
+    // bits
+    //                       3322222222221111 1111110000000000
+    //                       1098765432109876 5432109876543210
+    // accum0 u axis coordinate            xx xxxxxxxxxxxxxxxx  18 bits
+    // after shift and mask                                 xx
+    // accum1 v axis                       xx xxxxxxxxxxxxxxxx  18 bits
+    // after shift and mask                               xx
+
+    // add_raw make the interpolator increment the accumulator
+    // with the base value without masking or shifting
+    let config0 = LaneCtrl {
+        add_raw: true,
+        shift: uv_fractional_bits,
+        mask_lsb: 0,
+        mask_msb: texture_width_bits - 1,
+        ..LaneCtrl::DEFAULT
+    };
+    interp.get_lane0().set_ctrl(config0.encode());
+    let config1 = LaneCtrl {
+        add_raw: true,
+        shift: uv_fractional_bits - texture_width_bits,
+        mask_lsb: texture_width_bits,
+        mask_msb: texture_width_bits + texture_height_bits - 1,
+        ..LaneCtrl::DEFAULT
+    };
+    interp.get_lane1().set_ctrl(config1.encode());
+
+    interp.set_base(0);
+
+    // set starting position to 0x0
+    // will move 1/2 a pixel horizontally
+    // and 1/3 a pixel vertically per call to pop()
+    interp.get_lane0().set_accum(0);
+    interp.get_lane0().set_base(65536 / 2);
+    interp.get_lane1().set_accum(0);
+    interp.get_lane1().set_base(65536 / 3);
+
+    let expected = [
+        0x00, 0x00, 0x01, 0x01, 0x12, 0x12, 0x13, 0x23, 0x20, 0x20, 0x31, 0x31,
+    ];
+
+    for i in expected {
+        if i != texture[interp.pop() as usize] {
+            return false;
+        }
+    }
+
+    // reset the starting position
+    interp.get_lane0().set_accum(0);
+    interp.get_lane1().set_accum(0);
+    interp.set_base(texture.as_ptr() as u32);
+
+    for i in expected {
+        // This is unsafe and should be done extremely carefully
+        // remember to follow memory alignment,
+        // reading or writing an unaligned address will crash
+        if i != unsafe { *(interp.pop() as *const u8) } {
```

Comment:
interp.pop() always returns u32, right?
does it make sense to use raw pointers here, when `interp.pop() as u8` or `(interp.pop & 0xFF) as u8`(if you want to be explicit) would be equivalent?
if you want to do a u8 volatile read at that address as the C example does you'll have to drop down to pac level (or add a pop_u8 variant)

Reply:
It is to be closest to the example in c++
it reads u32 from the pop register
which I cast into a pointer to u8.
the u8 read is from the array texture: [u8;16]

Reply:
Oh... wow, I totally misread what this was doing.
Thanks for clarifying, this makes sense now.
I think I might have a go at documenting this in the future, since I doubt I'll be the only one to get confused by this,